File: unid-node-wallet-sdk/src/did-unid/keyring/secp256k1.ts
import base64url from 'base64url'
import { KeyPairSecp256K1, PublicKeyPayload } from '../../core'
import { utils } from '../../utils/utils'
import { Runtime } from '../../runtime'

interface Secp256k1Context {
    public : Buffer,
    private: Buffer,
}

export interface Secp256k1HexKeyPair {
    public : string,
    private: string,
}

const PRIVATE_KEY_SIZE            : number = 32 // Buffer(PrivateKey (32 = 256 bit))
const COMPRESSED_PUBLIC_KEY_SIZE  : number = 33 // Buffer(0x04 + PublicKey (32 = 256 bit))
const UNCOMPRESSED_PUBLIC_KEY_SIZE: number = 65 // Buffer(0x04 + PublicKey (64 = 512 bit))

export class Secp256k1 {
    private _public : Buffer
    private _private: Buffer

    /**
     * @param context 
     */
    constructor(context: Secp256k1Context) {
        if (context.private.length !== PRIVATE_KEY_SIZE) {
            throw new Error()
        }
        this._private = context.private

        switch (context.public.length) {
            case COMPRESSED_PUBLIC_KEY_SIZE: {
                this._public = this.transformUncompressedPublicKey(context.public)
                break
            }
            case UNCOMPRESSED_PUBLIC_KEY_SIZE: {
                this._public = context.public
                break
            }
            default: {
                throw new Error()
            }
        }
    }

    /**
     * @returns
     */
    public getPublicKey(): Buffer {
        return this._public
    }

    /**
     * @returns
     */
    public getPrivateKey(): Buffer {
        return this._private
    }

    /**
     * @returns
     */
    public toHexKeyPair(): Secp256k1HexKeyPair {
        return {
            public : this.getPublicKey().toString('hex'),
            private: this.getPrivateKey().toString('hex'),
        }
    }

    /**
     * @param jwk 
     * @returns
     */
    public static fromJwk(jwk: KeyPairSecp256K1): Secp256k1 {
        let d = jwk.d
        let x = jwk.x
        let y = jwk.y

        if (d === undefined) {
            d = base64url.encode(
                Buffer.from(utils.range(0, PRIVATE_KEY_SIZE).map(() => {
                    return 0x0
                }))
            )
        }

        return new Secp256k1({
            private: Buffer.from(base64url.toBuffer(d)),
            public : Buffer.from(
                Buffer.concat([
                    Buffer.from(Uint8Array.from([ 0x04 ])),
                    Buffer.from(Uint8Array.from(base64url.toBuffer(x))),
                    Buffer.from(Uint8Array.from(base64url.toBuffer(y))),
                ])
            ),
        })
    }

    /**
     * @param includedPrivateKey 
     * @returns
     */
    public toJwk(includedPrivateKey: boolean = false): KeyPairSecp256K1 {
        if (! this.validatePoint()) {
            throw new Error()
        }

        const jwk: KeyPairSecp256K1 = {
            kty: 'EC',
            crv: 'secp256k1',
            x  : base64url.encode(this.getPointX()),
            y  : base64url.encode(this.getPointY()),
        }

        if (includedPrivateKey) {
            jwk.d = base64url.encode(this.getPrivateKey())
        }

        return jwk
    }

    /**
     * @param keyId 
     * @param purpose 
     * @returns
     */
    public toPublicKey(keyId: string, purpose: Array<string>): PublicKeyPayload {
        if (! this.validatePoint()) {
            throw new Error()
        }

        return {
            id  : keyId,
            type: 'EcdsaSecp256k1VerificationKey2019',
            jwk : this.toJwk(),
            purpose: purpose,
        }
    }

    /**
     * @returns
     */
    public getPointX(): Buffer {
        if (this.getPublicKey().length !== UNCOMPRESSED_PUBLIC_KEY_SIZE) {
            throw new Error()
        }
        if (this.getPublicKey()[0] !== 0x04) {
            throw new Error()
        }
        const n = this.getPublicKey().slice(1)
        const s = n.slice(0, 32)

        return s
    }
    
    /**
     * @returns
     */
    public getPointY(): Buffer {
        if (this.getPublicKey().length !== UNCOMPRESSED_PUBLIC_KEY_SIZE) {
            throw new Error()
        }
        if (this.getPublicKey()[0] !== 0x04) {
            throw new Error()
        }
        const n = this.getPublicKey().slice(1)
        const s = n.slice(32)

        return s
    }
    
    /**
     * @returns
     */
    public validatePoint(): boolean {
        const nx = parseInt(this.getPointX().toString('hex'), 16)
        const ny = parseInt(this.getPointY().toString('hex'), 16)
        const np = parseInt('FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F', 16)
    
        return ((ny * ny - nx * nx * nx - 7) % np === 0)
    }

    /**
     * @param compressed 
     * @returns
     */
    private transformUncompressedPublicKey(compressed: Buffer): Buffer {
        return Buffer.from(Runtime.Secp256k1.publicKeyConvert(Uint8Array.from(compressed), false))
    }
}

File: src/did_unid/secp256k1.rs
use super::interfaces::keyring_model::Secp256k1;
use crate::core::interfaces::did_document::*;
use crate::core::interfaces::did_operation::*;

pub struct Secp256k1Context {
    pub public: Vec<u8>,
    pub private: Vec<u8>,
}

pub struct Secp256k1HexKeyPair {
    pub public: String,
    pub private: String,
}

const PRIVATE_KEY_SIZE: usize = 32; // Buffer(PrivateKey (32 = 256 bit))
const COMPRESSED_PUBLIC_KEY_SIZE: usize = 33; // Buffer(0x04 + PublicKey (32 = 256 bit))
const UNCOMPRESSED_PUBLIC_KEY_SIZE: usize = 65; // Buffer(0x04 + PublicKey (64 = 512 bit))

impl Secp256k1 {
    pub fn new(context: Secp256k1Context) -> Self {
        let private: Vec<u8>;
        let public: Vec<u8>;
        assert_eq!(context.private.len(), PRIVATE_KEY_SIZE);
        private = context.private;
        match context.public.len() {
            COMPRESSED_PUBLIC_KEY_SIZE => {
                public = Secp256k1::transform_uncompressed_public_key(context.public);
            }
            UNCOMPRESSED_PUBLIC_KEY_SIZE => {
                public = context.public;
            }
            _ => panic!("public key length is invalid"),
        };
        Secp256k1 {
            _public: public,
            _private: private,
        }
    }

    pub fn get_point_x(&self) -> Vec<u8> {
        assert_eq!(self._public.len(), UNCOMPRESSED_PUBLIC_KEY_SIZE);
        assert_eq!(self._public[0], 0x04);
        let n: &[u8] = &self.get_public_key()[1..];
        let s: &[u8] = &n[0..32];
        s.to_vec()
    }

    pub fn get_point_y(&self) -> Vec<u8> {
        assert_eq!(self._public.len(), UNCOMPRESSED_PUBLIC_KEY_SIZE);
        assert_eq!(self._public[0], 0x04);
        let n: &[u8] = &self.get_public_key()[1..];
        let s: &[u8] = &n[32..];
        s.to_vec()
    }

    pub fn to_jwk(&self) -> KeyPairSecp256K1 {
        let jwk: KeyPairSecp256K1 = KeyPairSecp256K1 {
            kty: "EC".to_string(),
            crv: "secp256k1".to_string(),
            x: base64_url::encode(&self.get_point_x()[..]),
            y: base64_url::encode(&self.get_point_y()[..]),
        };
        jwk
    }

    pub fn get_public_key(&self) -> Vec<u8> {
        self._public.clone()
    }

    pub fn get_private_key(&self) -> Vec<u8> {
        self._private.clone()
    }

    pub fn to_public_key(&self, key_id: String, purpose: Vec<String>) -> PublicKeyPayload {
        PublicKeyPayload {
            id: key_id,
            type_field: "EcdsaSecp256k1VerificationKey2019".to_string(),
            jwk: self.to_jwk(),
            purpose,
        }
    }

    pub fn to_hex_key_pair(&self) -> Secp256k1HexKeyPair {
        Secp256k1HexKeyPair {
            public: hex::encode(self.get_public_key()),
            private: hex::encode(self.get_private_key()),
        }
    }

    fn transform_uncompressed_public_key(compressed: Vec<u8>) -> Vec<u8> {
        crate::runtime::secp256k1::Secp256k1::public_key_convert(compressed, false)
    }
}
